torch/csrc/jit/passes/onnx.cpp

```diff
@@ -186,11 +187,40 @@ void BlockToONNX(
     return it->second;
   };
 
+  GRAPH_DEBUG("BlockToONNX: graph of old block: ", old_block->owningGraph()->toString());
+
   // Initialize context and environment
   for (auto input : old_block->inputs()) {
     auto n = ctx.block->addInput()->copyMetadata(input);
     env[input] = n;
   }
+
+  // Merge ONNX inferred type with existing type.
+  // If inferred type is sequence, overwrite old type because it is more compatible
+  // with ONNX shape inference.
+  // If inferred type is tensor type, merge the two types, with inferred type having
+  // higher precedence.
+  auto mergeOutputTypes = [](TypePtr old_type, TypePtr new_type) -> TypePtr {
+    auto new_list_type = new_type->cast<ListType>();
```

Comment:
Is there a possible prelim check we can do right before this casting op to see if `new_type` is of one of the types that support casting to `ListType`? Also, would it make sense to extract this lambda out as a function so that it can be used elsewhere?

Reply:
casting and checking if nullptr seems to be the only way of checking if it is a certain underlying type. Moved this to a standalone function in `shape_type_inference.cpp`.